src/main/java/frc/robot/subsystems/DriveSubsystem.java

```diff
@@ -7,18 +7,104 @@
 
 package frc.robot.subsystems;
 
+import com.revrobotics.CANEncoder;
+
+import edu.wpi.first.wpilibj.ADXRS450_Gyro;
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.PWMVictorSPX;
+import edu.wpi.first.wpilibj.SpeedControllerGroup;
 import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj.geometry.Pose2d;
+import edu.wpi.first.wpilibj.interfaces.Gyro;
+import edu.wpi.first.wpilibj.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.wpilibj.kinematics.DifferentialDriveWheelSpeeds;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 
+
 public class DriveSubsystem extends SubsystemBase {
 
     private final DifferentialDrive driveTrain;
+    private final SpeedControllerGroup leftMotors;
+    private final SpeedControllerGroup rightMotors;
+    private final CANEncoder leftEncoder;
+    private final CANEncoder rightEncoder;
+
+    private final Gyro m_gyro = new ADXRS450_Gyro();
+    private final DifferentialDriveOdometry m_odometry;
+
 
-    public DriveSubsystem(DifferentialDrive driveTrain) {
+    public DriveSubsystem(DifferentialDrive driveTrain, SpeedControllerGroup leftMotors, SpeedControllerGroup rightMotors, CANEncoder leftEncoder, CANEncoder rightEncoder) {
         this.driveTrain = driveTrain;
+        this.leftMotors = leftMotors;
+        this.rightMotors = rightMotors;
+        this.leftEncoder = leftEncoder;
+        this.rightEncoder = rightEncoder;
+
+        resetEncoders();
+        m_odometry = new DifferentialDriveOdometry(m_gyro.getRotation2d());
     }
 
     public void drive(double speed, double rotation) {
         driveTrain.arcadeDrive(speed, rotation);
     }
+
+    @Override
+    public void periodic() {
+    // Update the odometry in the periodic block
+        m_odometry.update(m_gyro.getRotation2d(), leftEncoder.getPosition(),
+                      rightEncoder.getPosition());
+    }
+
+
+    public void tankDriveVolts(double leftVolts, double rightVolts) {
+        leftMotors.setVoltage(leftVolts);
+        rightMotors.setVoltage(-rightVolts);
+        driveTrain.feed();
```

Comment:
We're using `setVoltage` here since the drivetrain does only uses `set` under the hood. `setVoltage` scales with the battery's current voltage, so I think that's why it's used here.

Reply:
When would you ever need this?